docs/_source/tutorials/notebooks/training-llm-mistral-sft.ipynb

```diff
@@ -0,0 +1,602 @@
+    "model_id = \"mistralai/Mistral-7B-v0.1\"\n",
+    "model = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float16, device_map=\"auto\")\n",
+    "# tokenizer = AutoTokenizer.from_pretrained(model_id, padding_side='left', add_eos_token=True)\n",
+    "# tokenizer_with_eos = AutoTokenizer.from_pretrained(model_id)#, padding_side='right')\n",
+    "# tokenizer_with_eos.pad_token_id = tokenizer_with_eos.eos_token_id\n",
+    "# tokenizer_with_eos._tokenizer.post_processor = TemplateProcessing(\n",
+    "#     single=f\"$A {tokenizer_with_eos.eos_token}\",\n",
+    "#     special_tokens=[(tokenizer_with_eos.eos_token, tokenizer_with_eos.eos_token_id)],\n",
+    "# )\n",
+    "tokenizer = AutoTokenizer.from_pretrained(model_id)\n",
+    "tokenizer.pad_token_id = tokenizer.eos_token_id\n",
+    "data_collator = DataCollatorForSeq2SeqCopyLabels(tokenizer=tokenizer, model=model)\n",
+    "\n",
+    "\"\"\"\n",
+    "tokenizer_without_eos = AutoTokenizer.from_pretrained(model_id)\n",
+    "tokenizer_without_eos.pad_token_id = tokenizer_without_eos.eos_token_id\n",
+    "\"\"\"\n",
+    "\n",
+    "def formatting_func(sample: Dict[str, Any]) -> Iterator[str]:\n",
+    "    if len(sample[\"response\"]) <= 2:\n",
+    "        return\n",
+    "    prompt = sample[\"prompt\"]\n",
+    "    prompt = prompt.replace(\"<human>:\", f\"{tokenizer.eos_token}{tokenizer.bos_token}<human>:\")\n",
+    "    prompt = prompt[prompt.find(\"<human>:\"):]\n",
+    "    output = prompt + \" \" + sample[\"response\"]\n",
+    "    if sample[\"background\"]:\n",
+    "        output = sample[\"background\"] + \" \" + output\n",
+    "    output = output + \"</s>\"\n",
+    "    assert output.count(\"<s>\") + 1 == output.count(\"</s>\")\n",
+    "    return output\n",
+    "\n",
+    "task = TrainingTask.for_supervised_fine_tuning(formatting_func)\n",
+    "formatted_dataset = dataset.prepare_for_training(framework=\"trl\", task=task)\n",
+    "\n",
+    "collated = data_collator([tokenizer(formatted_dataset[3][\"text\"])])\n",
+    "print(collated)\n",
+    "print(tokenizer.decode(collated[\"input_ids\"][0]))\n",
+    "# breakpoint()\n",
+    "\n",
+    "\"\"\"\n",
+    "import torch\n",
+    "from transformers import StoppingCriteria\n",
+    "\n",
+    "class StoppingCriteriaTokens(StoppingCriteria):\n",
+    "    def __init__(self, stopping_input_ids = None):\n",
+    "        super().__init__()\n",
+    "        self.stopping_input_ids = stopping_input_ids\n",
+    "\n",
+    "    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> bool:\n",
+    "        for stopping_input_ids in self.stopping_input_ids:\n",
+    "            length = len(stopping_input_ids)\n",
+    "            if input_ids[0, -length:].tolist() == stopping_input_ids:\n",
+    "                return True\n",
+    "\n",
+    "stopping_criteria = StoppingCriteriaList([StoppingCriteriaTokens(tokenizer_without_eos([\"<bot>:\", \"<human>:\"]).input_ids)])\n",
```

Comment:
I would highlight the stopping criteria too

Reply:
This has also been removed - `StoppingCriteriaList` was too inconsistent of an approach to stop with, the current way with EOS is much more consistent. This was just leftover from an earlier version